### PR TITLE
TestNg - Make Display disabled tests in report configurable (fixes #369)

### LINKED ISSUE #369: [TestNG] Make "Display disabled tests in report" configurable
#### I'm submitting a ... 
  - [ ] bug report
  - [x] feature request
  - [ ] support request => Please do not submit support request here, see note at the top of this template.

#### What is the current behavior?

Disabled tests (`@Test(enabled = false`)) appear as unknown in the report.

https://github.com/allure-framework/allure-java/pull/275

#### What is the expected behavior?

Make it configurable so there would be an option to allow disabled tests to appear in report or not.

#### What is the motivation / use case for changing the behavior?

We have a test suite which supports the same application with different features in multiple countries. Some countries do not have features that other countries have thus we're using Annotation Transformer to disable those tests for those countries. 

It would be nice to not see those disabled tests in the report as they clutter the report and create a distraction.

#### Please tell us about your environment:

| Allure version | 2.8 |
| --- | --- |
| Test framework | testng@6.14.3 |
| Allure integration | - |
| Generate report using | allure-gradle@2.5 |

We are not updating to the newest versions because of that behavior and we would love to avoid forking and changing this behavior in forked Allure projects.

### PR DESCRIPTION
**Pull request by vinipx**
### Context

#369

#### Checklist
- [x] [Sign Allure CLA][cla]
- [x] Provide unit tests

[cla]: https://cla-assistant.io/accept/allure-framework/allure2

### CODE REVIEW

File: allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import java.util.Properties;
+
+public class AllureTestNgConfig {
+
+    public static final String ALLURE_TESTNG_CONFIG = "allure.testng.config.";
+    public static final String ALLURE_REPORT_DISABLED_TESTS = ALLURE_TESTNG_CONFIG + "report.disabled.tests";
+    private boolean disabledTestsReported;
+
+    public AllureTestNgConfig() {
+        final Properties properties = PropertiesUtils.loadAllureProperties();
```
Comment (baev) [inline allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java:14]: lets keep this POJO as simple as possible. Move initialisation to static factory method

File: allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        disabledTestsReported = Boolean.parseBoolean(properties.getProperty(ALLURE_REPORT_DISABLED_TESTS));
```
Comment (baev) [inline allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java:15]: lets keep default behaviour -- disabled tests are reported when property is omitted

File: allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+import io.qameta.allure.util.PropertiesUtils;
+
+import java.util.Properties;
+
+public class AllureTestNgConfig {
+
+    public static final String ALLURE_TESTNG_CONFIG = "allure.testng.config.";
+    public static final String ALLURE_REPORT_DISABLED_TESTS = ALLURE_TESTNG_CONFIG + "report.disabled.tests";
+    private boolean disabledTestsReported;
```
Comment (baev) [inline allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java:11]: rename it to `hideDisabledTests`

File: allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package io.qameta.allure.testng.config;
+
+import io.qameta.allure.util.PropertiesUtils;
+
+import java.util.Properties;
+
+public class AllureTestNgConfig {
+
+    public static final String ALLURE_TESTNG_CONFIG = "allure.testng.config.";
```
Comment (baev) [inline allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java:9]: omit the `config.` suffix as unnecessary

File: allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java [hunk 5]
```diff
@@ -1055,19 +1057,38 @@ public void shouldProcessConfigurationFailure() {
     }
 
     @AllureFeatures.IgnoredTests
-    @Issue("49")
+    @Issue("369")
```
Comment (baev) [inline allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java:1060]: why do you change the issue, please revert. Default behaviour shouldn't be changed, so simply add a new test

File: allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java [hunk 6]
```diff
@@ -1431,6 +1452,7 @@ public void idAssignToOtherTest() {
     @Test
     @AllureFeatures.Filtration
     public void skippedTest() {
+        System.setProperty(ALLURE_REPORT_DISABLED_TESTS, "true");
```
Comment (baev) [inline allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java:1455]: make sure you cleanup the system property. I'm aware of `SetSystemProperty` JUnit5 extension, but not sure how it supposed to be done in TestNG.

Maybe we can simply provide config options directly to config POJO and test AllureTestNgConfig initialisation in separate test

**Review by baev — CHANGES_REQUESTED** [review]

File: allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java (hunk 1 above)
Comment (vinipx) [reply allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java]: Done, pls check.

File: allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java (hunk 2 above)
Comment (vinipx) [reply allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java]: Agree - fixed by making default behavior to report tests when property is omitted/false.

File: allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java (hunk 4 above)
Comment (vinipx) [reply allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java]: Cool, I've changed it to "allure.testng.hide.disabled.tests"

File: allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java (hunk 5 above)
Comment (vinipx) [reply allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java]: Agree - reverted back.

File: allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java (hunk 6 above)
Comment (vinipx) [reply allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java]: Done, thanks. Please, review.

**Comment by baev** [comment]
@vinipx could you please fix the build?

**Comment by vinipx** [comment]
> @vinipx could you please fix the build?

Done, thanks.

File: allure-testng/src/test/resources/allure.properties [hunk 7]
```diff
@@ -1,3 +1,4 @@
 allure.results.directory=build/allure-results
 allure.link.issue.pattern=https://github.com/allure-framework/allure-java/issues/{}
 allure.label.epic=#project.description#
+allure.testng.hide.disabled.tests=false
```
Comment (baev) [inline allure-testng/src/test/resources/allure.properties:4]: remove

**Review by baev — CHANGES_REQUESTED** [review]

File: allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java [hunk 8]
```diff
@@ -72,6 +75,8 @@
 @SuppressWarnings("deprecation")
 public class AllureTestNgTest {
 
+    private static AllureTestNgConfig ALLURE_CONFIGURATION = AllureTestNgConfig.loadConfigProperties();
```
[the remark this answers is not in the document]
Comment (vinipx) [reply allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java]: done

File: allure-testng/src/test/resources/allure.properties (hunk 7 above)
Comment (vinipx) [reply allure-testng/src/test/resources/allure.properties]: done

File: allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java [hunk 9]
```diff
@@ -1063,11 +1074,23 @@ public void shouldDisplayDisabledTests() {
         assertThat(results.getTestResults())
                 .extracting(TestResult::getName, TestResult::getStatus)
                 .containsExactly(
```
Comment (baev) [inline allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java:1076]: containsExactlyInAnyOrder

File: allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java [hunk 10]
```diff
@@ -1063,11 +1074,23 @@ public void shouldDisplayDisabledTests() {
...
     }
 
+    @AllureFeatures.IgnoredTests
+    @Issue("369")
+    @Test
+    public void shouldNotDisplayDisabledTests() {
+        allureTestNgConfig.setConfiguration(ALLURE_TESTNG_HIDE_DISABLED_TESTS, "true");
+        final AllureResults results = runTestNgSuites("suites/gh-369.xml");
+
+        assertThat(results.getTestResults()).extracting(TestResult::getName, TestResult::getStatus).containsOnly(tuple("enabled", Status.PASSED));
```
Comment (baev) [inline allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java:1090]: move each chaining method call to a new line to improve the readability

File: allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.util.Properties;
+
+import static java.lang.Boolean.parseBoolean;
+import static java.lang.System.setProperty;
+
+public class AllureTestNgConfig {
+
+    public static final String ALLURE_TESTNG_HIDE_DISABLED_TESTS = "allure.testng.hide.disabled.tests";
+    private static boolean hideDisabledTests;
```
Comment (baev) [inline allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java:28]: why this one is static?

File: allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+    public boolean isHideDisabledTests() {
+        return hideDisabledTests;
+    }
+
+    public AllureTestNgConfig setConfiguration(final String property, final String value) {
```
Comment (baev) [inline allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java:34]: lets do `setHideDisabledTests` instead

File: allure-testng/src/main/java/io/qameta/allure/testng/AllureTestNg.java [hunk 13]
```diff
@@ -200,11 +204,13 @@ public void onStart(final ITestContext context) {
                 .distinct()
                 .forEach(this::onBeforeClass);
 
-        context.getExcludedMethods().stream()
-                .filter(ITestNGMethod::isTest)
-                .filter(method -> !method.getEnabled())
-                .filter(testFilter::isSelected)
-                .forEach(method -> createFakeResult(context, method));
+        if (config.isHideDisabledTests()) {
```
Comment (baev) [inline allure-testng/src/main/java/io/qameta/allure/testng/AllureTestNg.java:207]: that should be `!config.isHideDisabledTests()`. And why the tests pass?

**Review by baev — CHANGES_REQUESTED** [review]

File: allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java (hunk 9 above)
Comment (vinipx) [reply allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java]: good catch, thanks. Done.

File: allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java (hunk 11 above)
Comment (vinipx) [reply allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java:28]: If we make it non-static, then **loadConfigProperties()** needs to be non-static.
Consequently, we might need to instantiate AllureTestNgConfig in AllureTestNg constructor or do similar thing - not sure if makes sense though:

line: 148 
_this.config = new AllureTestNgConfig.loadConfigProperties();_

What are your thoughts?

File: allure-testng/src/main/java/io/qameta/allure/testng/AllureTestNg.java (hunk 13 above)
Comment (baev) [reply allure-testng/src/main/java/io/qameta/allure/testng/AllureTestNg.java:207]: as far as I can see the block actually adds disabled tests to the report. If false condition, disabled tests are not reported

File: allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java (hunk 11 above)
Comment (baev) [reply allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java:28]: AllureTestNgConfig should be a simple POJO that accept all the configuration properties in constructor. AllureTestNgConfig.loadConfigProperties then is a simple factory method that loads Allure properties and passes values to constructor.

File: allure-testng/src/main/java/io/qameta/allure/testng/AllureTestNg.java (hunk 13 above)
Comment (vinipx) [reply allure-testng/src/main/java/io/qameta/allure/testng/AllureTestNg.java]: Correct - my mistake. 
Fixed, thanks.

File: allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java (hunk 11 above)
Comment (vinipx) [reply allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java:28]: Cool, I think I managed to implement following your suggestion. Please, let me know if it makes sense

File: allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.Properties;
+
+import static java.lang.Boolean.parseBoolean;
+
+public class AllureTestNgConfig {
+
+    public static final String ALLURE_TESTNG_HIDE_DISABLED_TESTS = "allure.testng.hide.disabled.tests";
+    private boolean hideDisabledTests;
+
+    public AllureTestNgConfig(Properties properties) {
```
Comment (baev) [inline allure-testng/src/main/java/io/qameta/allure/testng/config/AllureTestNgConfig.java:29]:
```suggestion
    public AllureTestNgConfig(final Properties properties) {
```

File: allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java [hunk 15]
```diff
@@ -72,6 +74,8 @@
 @SuppressWarnings("deprecation")
 public class AllureTestNgTest {
 
+    //private AllureTestNgConfig allureTestNgConfig = AllureTestNgConfig.loadConfigProperties();
```
Comment (baev) [inline allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java:77]:
```suggestion
```

**Review by baev — APPROVED** [review]

**Comment by baev** [comment]
@vinipx thanks!

**Pull request merged**